pkg/agent/supportbundlecollection/support_bundle_controller.go

```diff
@@ -0,0 +1,439 @@
+		arr := strings.Split(supportBundle.FileServer.URL, "://")
+		if strings.ToLower(arr[0]) != string(sftpProtocol) {
+			return fmt.Errorf("only sftp is supported for now, protocol %s is not supported", strings.ToUpper(arr[0]))
+		}
+		url = supportBundle.FileServer.URL[7:]
+	} else {
+		url = supportBundle.FileServer.URL
+	}
+
+	triesLeft := uploadToFileServerTries
+	var uploadErr error
+	for triesLeft > 0 {
+		if uploadErr = c.uploadToFileServer(supportBundle.Name, url, &supportBundle.Authentication, outputFile); uploadErr == nil {
+			return nil
+		}
+		triesLeft -= 1
+		if triesLeft == 0 {
+			return fmt.Errorf("aborting upload support bundle after %d attempts", uploadToFileServerTries)
+		}
+		klog.InfoS("Failed to upload support bundle", "UploadError", uploadErr, "TriesLeft", triesLeft)
+		time.Sleep(uploadToFileServerRetryDelay)
+	}
+	return nil
+}
+
+func (c *SupportBundleController) uploadToFileServer(bundleName string, url string, serverAuth *cpv1b2.BundleServerAuthConfiguration, tarGzFile io.Reader) error {
+	strArr := strings.Split(url, "/")
+	address := strArr[0]
+	targetPath := url[len(address):]
+	path := path.Join(targetPath, c.nodeName+"_"+bundleName+".tar.gz")
+	cfg := &ssh.ClientConfig{
+		User: serverAuth.BasicAuthentication.Username,
+		Auth: []ssh.AuthMethod{ssh.Password(serverAuth.BasicAuthentication.Password)},
+		// #nosec G106: skip host key check here and users can specify their own checks if needed
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+		Timeout:         time.Second,
+	}
+	uploader, err := getUploader(sftpProtocol)
+	if err != nil {
+		return err
+	}
+	return uploader.upload(address, path, cfg, tarGzFile)
+}
+
+func getUploaderByProtocol(protocol ProtocolType) (uploader, error) {
+	if protocol == sftpProtocol {
+		if sftpUploaderInstance == nil {
+			sftpUploaderInstance = &sftpUploader{}
+		}
+		return sftpUploaderInstance, nil
+	}
+	return nil, fmt.Errorf("unsupported protocol %s", protocol)
+}
+
+type uploader interface {
+	upload(addr string, path string, config *ssh.ClientConfig, tarGzFile io.Reader) error
+}
+
+type sftpUploader struct {
+}
+
+func (uploader *sftpUploader) upload(address string, path string, config *ssh.ClientConfig, tarGzFile io.Reader) error {
+	conn, err := ssh.Dial("tcp", address, config)
+	if err != nil {
+		return fmt.Errorf("error when connecting to fs server: %w", err)
+	}
+	sftpClient, err := sftp.NewClient(conn)
+	if err != nil {
+		return fmt.Errorf("error when setting up sftp client: %w", err)
+	}
+	defer func() {
+		if err := sftpClient.Close(); err != nil {
+			klog.Errorf("Error when closing sftp client: %v", err)
+		}
+	}()
+	targetFile, err := sftpClient.Create(path)
+	if err != nil {
+		return fmt.Errorf("error when creating target file on remote: %v", err)
+	}
+	defer func() {
+		if err := targetFile.Close(); err != nil {
+			klog.Errorf("error when closing target file on remote: %v", err)
+		}
+	}()
+	if written, err := io.Copy(targetFile, tarGzFile); err != nil {
```

Comment:
Question: if uploading file failed in the first time, will it leave stale files on the remote server and break the second try?

Reply:
I don't think it will break the second try and it should replace the old one with the latest one.